File: python/Textract/src/textract_processor.py
"""
Direct AWS Textract processor for document analysis
"""
import boto3
import json
from typing import Dict, Any, List, Optional
from botocore.exceptions import ClientError

class TextractProcessor:
    def __init__(self, region_name: str = 'us-east-1'):
        self.textract_client = boto3.client('textract', region_name=region_name)
        self.s3_client = boto3.client('s3', region_name=region_name)
    
    def process_document_sync(self, document_bytes: bytes, doc_type: str) -> Dict[str, Any]:
        """Process document synchronously using Textract"""
        try:
            response = self.textract_client.detect_document_text(
                Document={'Bytes': document_bytes}
            )
            
            # Also get forms and tables
            forms_response = self.textract_client.analyze_document(
                Document={'Bytes': document_bytes},
                FeatureTypes=['FORMS', 'TABLES']
            )
            
            return self._parse_textract_response(response, forms_response, doc_type)
            
        except ClientError as e:
            raise Exception(f"Textract processing error: {str(e)}")
    
    def process_document_async(self, s3_bucket: str, s3_key: str, doc_type: str) -> str:
        """Start asynchronous document processing"""
        try:
            response = self.textract_client.start_document_analysis(
                DocumentLocation={
                    'S3Object': {
                        'Bucket': s3_bucket,
                        'Name': s3_key
                    }
                },
                FeatureTypes=['FORMS', 'TABLES']
            )
            
            return response['JobId']
            
        except ClientError as e:
            raise Exception(f"Textract async processing error: {str(e)}")
    
    def get_async_results(self, job_id: str, doc_type: str) -> Dict[str, Any]:
        """Get results from asynchronous processing"""
        try:
            response = self.textract_client.get_document_analysis(JobId=job_id)
            
            if response['JobStatus'] == 'SUCCEEDED':
                return self._parse_textract_response(None, response, doc_type)
            elif response['JobStatus'] == 'FAILED':
                raise Exception(f"Textract job failed: {response.get('StatusMessage', 'Unknown error')}")
            else:
                return {'status': response['JobStatus'], 'job_id': job_id}
                
        except ClientError as e:
            raise Exception(f"Error getting async results: {str(e)}")
    
    def _parse_textract_response(self, text_response: Optional[Dict], 
                                forms_response: Dict, doc_type: str) -> Dict[str, Any]:
        """Parse Textract response based on document type"""
        
        if doc_type == 'w2':
            return self._parse_w2_document(text_response, forms_response)
        elif doc_type == 'bank_statement':
            return self._parse_bank_statement(text_response, forms_response)
        
        return {'raw_response': forms_response}
    
    def _parse_w2_document(self, text_response: Optional[Dict], 
                          forms_response: Dict) -> Dict[str, Any]:
        """Parse W-2 specific fields from Textract response"""
        w2_fields = {
            'employee_ssn': None,
            'employee_name': None,
            'employer_ein': None,
            'employer_name': None,
            'wages': None,
            'federal_tax_withheld': None,
            'social_security_wages': None,
            'medicare_wages': None
        }
        
        blocks = forms_response.get('Blocks', [])
        key_value_pairs = self._extract_key_value_pairs(blocks)
        
        # Map Textract fields to W-2 fields
        field_mappings = {
            'Employee\'s social security number': 'employee_ssn',
            'Employee\'s name': 'employee_name',
            'Employer identification number': 'employer_ein',
            'Employer\'s name': 'employer_name',
            'Wages, tips, other compensation': 'wages',
            'Federal income tax withheld': 'federal_tax_withheld'
        }
        
        for key, value in key_value_pairs.items():
            for mapping_key, field_name in field_mappings.items():
                if mapping_key.lower() in key.lower():
                    w2_fields[field_name] = value
                    break
        
        return {
            'document_type': 'w2',
            'extracted_fields': w2_fields,
            'confidence_scores': self._calculate_confidence_scores(blocks),
            'raw_key_value_pairs': key_value_pairs
        }
    
    def _parse_bank_statement(self, text_response: Optional[Dict], 
                             forms_response: Dict) -> Dict[str, Any]:
        """Parse bank statement specific fields from Textract response"""
        statement_data = {
            'account_number': None,
            'account_holder': None,
            'statement_period': None,
            'beginning_balance': None,
            'ending_balance': None,
            'transactions': []
        }
        
        blocks = forms_response.get('Blocks', [])
        
        # Extract tables (transactions)
        tables = self._extract_tables(blocks)
        statement_data['transactions'] = self._parse_transaction_tables(tables)
        
        # Extract key information
        key_value_pairs = self._extract_key_value_pairs(blocks)
        
        return {
            'document_type': 'bank_statement',
            'extracted_fields': statement_data,
            'confidence_scores': self._calculate_confidence_scores(blocks),
            'raw_key_value_pairs': key_value_pairs
        }
    
    def _extract_key_value_pairs(self, blocks: List[Dict]) -> Dict[str, str]:
        """Extract key-value pairs from Textract blocks"""
        key_value_pairs = {}
        
        for block in blocks:
            if block['BlockType'] == 'KEY_VALUE_SET':
                if 'KEY' in block.get('EntityTypes', []):
                    # This is a key block
                    key_text = self._get_text_from_relationships(block, blocks)
                    value_text = self._get_value_from_key_block(block, blocks)
                    if key_text and value_text:
                        key_value_pairs[key_text] = value_text
        
        return key_value_pairs
    
    def _extract_tables(self, blocks: List[Dict]) -> List[Dict]:
        """Extract table data from Textract blocks"""
        tables = []
        
        for block in blocks:
            if block['BlockType'] == 'TABLE':
                table_data = self._parse_table_block(block, blocks)
                tables.append(table_data)
        
        return tables
    
    def _get_text_from_relationships(self, block: Dict, all_blocks: List[Dict]) -> str:
        """Get text content from block relationships"""
        text_parts = []
        
        if 'Relationships' in block:
            for relationship in block['Relationships']:
                if relationship['Type'] == 'CHILD':
                    for child_id in relationship['Ids']:
                        child_block = next((b for b in all_blocks if b['Id'] == child_id), None)
                        if child_block and child_block['BlockType'] == 'WORD':
                            text_parts.append(child_block['Text'])
        
        return ' '.join(text_parts)
    
    def _calculate_confidence_scores(self, blocks: List[Dict]) -> Dict[str, float]:
        """Calculate confidence scores for extracted data"""
        confidences = []
        
        for block in blocks:
            if 'Confidence' in block:
                confidences.append(block['Confidence'])
        
        if confidences:
            return {
                'average': sum(confidences) / len(confidences),
                'minimum': min(confidences),
                'maximum': max(confidences)
            }
        
        return {'average': 0.0, 'minimum': 0.0, 'maximum': 0.0}

File: csharp/Textract/Services/TextractProcessor.cs
using Amazon.Textract;
using Amazon.Textract.Model;
using Amazon.S3;
using Textract.Models;

namespace Textract.Services
{
    public class TextractProcessor : ITextractProcessor
    {
        private readonly IAmazonTextract _textractClient;
        private readonly IAmazonS3 _s3Client;

        public TextractProcessor(IAmazonTextract textractClient, IAmazonS3 s3Client)
        {
            _textractClient = textractClient;
            _s3Client = s3Client;
        }

        public async Task<DocumentProcessingResult> ProcessDocumentSyncAsync(byte[] documentBytes, string docType)
        {
            try
            {
                // Detect text
                var textRequest = new DetectDocumentTextRequest
                {
                    Document = new Document
                    {
                        Bytes = new MemoryStream(documentBytes)
                    }
                };

                var textResponse = await _textractClient.DetectDocumentTextAsync(textRequest);

                // Analyze forms and tables
                var analyzeRequest = new AnalyzeDocumentRequest
                {
                    Document = new Document
                    {
                        Bytes = new MemoryStream(documentBytes)
                    },
                    FeatureTypes = new List<string> { "FORMS", "TABLES" }
                };

                var analyzeResponse = await _textractClient.AnalyzeDocumentAsync(analyzeRequest);

                return ParseTextractResponse(textResponse, analyzeResponse, docType);
            }
            catch (Exception ex)
            {
                throw new InvalidOperationException($"Textract processing error: {ex.Message}", ex);
            }
        }

        public async Task<string> ProcessDocumentAsync(string s3Bucket, string s3Key, string docType)
        {
            try
            {
                var request = new StartDocumentAnalysisRequest
                {
                    DocumentLocation = new DocumentLocation
                    {
                        S3Object = new S3Object
                        {
                            Bucket = s3Bucket,
                            Name = s3Key
                        }
                    },
                    FeatureTypes = new List<string> { "FORMS", "TABLES" }
                };

                var response = await _textractClient.StartDocumentAnalysisAsync(request);
                return response.JobId;
            }
            catch (Exception ex)
            {
                throw new InvalidOperationException($"Textract async processing error: {ex.Message}", ex);
            }
        }

        public async Task<DocumentProcessingResult> GetAsyncResultsAsync(string jobId, string docType)
        {
            try
            {
                var request = new GetDocumentAnalysisRequest
                {
                    JobId = jobId
                };

                var response = await _textractClient.GetDocumentAnalysisAsync(request);

                return response.JobStatus switch
                {
                    JobStatus.SUCCEEDED => ParseTextractResponse(null, response, docType),
                    JobStatus.FAILED => throw new InvalidOperationException($"Textract job failed: {response.StatusMessage}"),
                    _ => new DocumentProcessingResult
                    {
                        Status = response.JobStatus.Value,
                        JobId = jobId
                    }
                };
            }
            catch (Exception ex)
            {
                throw new InvalidOperationException($"Error getting async results: {ex.Message}", ex);
            }
        }

        private DocumentProcessingResult ParseTextractResponse(
            DetectDocumentTextResponse? textResponse,
            AnalyzeDocumentResponse analyzeResponse,
            string docType)
        {
            return docType switch
            {
                "w2" => ParseW2Document(textResponse, analyzeResponse),
                "bank_statement" => ParseBankStatement(textResponse, analyzeResponse),
                _ => new DocumentProcessingResult { RawResponse = analyzeResponse }
            };
        }

        private DocumentProcessingResult ParseW2Document(
            DetectDocumentTextResponse? textResponse,
            AnalyzeDocumentResponse analyzeResponse)
        {
            var w2Fields = new Dictionary<string, string?>
            {
                ["employee_ssn"] = null,
                ["employee_name"] = null,
                ["employer_ein"] = null,
                ["employer_name"] = null,
                ["wages"] = null,
                ["federal_tax_withheld"] = null,
                ["social_security_wages"] = null,
                ["medicare_wages"] = null
            };

            var blocks = analyzeResponse.Blocks;
            var keyValuePairs = ExtractKeyValuePairs(blocks);

            // Map Textract fields to W-2 fields
            var fieldMappings = new Dictionary<string, string>
            {
                ["Employee's social security number"] = "employee_ssn",
                ["Employee's name"] = "employee_name",
                ["Employer identification number"] = "employer_ein",
                ["Employer's name"] = "employer_name",
                ["Wages, tips, other compensation"] = "wages",
                ["Federal income tax withheld"] = "federal_tax_withheld"
            };

            foreach (var (key, value) in keyValuePairs)
            {
                foreach (var (mappingKey, fieldName) in fieldMappings)
                {
                    if (key.Contains(mappingKey, StringComparison.OrdinalIgnoreCase))
                    {
                        w2Fields[fieldName] = value;
                        break;
                    }
                }
            }

            return new DocumentProcessingResult
            {
                DocumentType = "w2",
                ExtractedFields = w2Fields,
                ConfidenceScores = CalculateConfidenceScores(blocks),
                RawKeyValuePairs = keyValuePairs
            };
        }

        private DocumentProcessingResult ParseBankStatement(
            DetectDocumentTextResponse? textResponse,
            AnalyzeDocumentResponse analyzeResponse)
        {
            var statementData = new Dictionary<string, object?>
            {
                ["account_number"] = null,
                ["account_holder"] = null,
                ["statement_period"] = null,
                ["beginning_balance"] = null,
                ["ending_balance"] = null,
                ["transactions"] = new List<Dictionary<string, object>>()
            };

            var blocks = analyzeResponse.Blocks;

            // Extract tables (transactions)
            var tables = ExtractTables(blocks);
            statementData["transactions"] = ParseTransactionTables(tables);

            // Extract key information
            var keyValuePairs = ExtractKeyValuePairs(blocks);

            return new DocumentProcessingResult
            {
                DocumentType = "bank_statement",
                ExtractedFields = statementData,
                ConfidenceScores = CalculateConfidenceScores(blocks),
                RawKeyValuePairs = keyValuePairs
            };
        }

        private Dictionary<string, string> ExtractKeyValuePairs(List<Block> blocks)
        {
            var keyValuePairs = new Dictionary<string, string>();

            foreach (var block in blocks.Where(b => b.BlockType == BlockType.KEY_VALUE_SET))
            {
                if (block.EntityTypes?.Contains("KEY") == true)
                {
                    var keyText = GetTextFromRelationships(block, blocks);
                    var valueText = GetValueFromKeyBlock(block, blocks);
                    
                    if (!string.IsNullOrEmpty(keyText) && !string.IsNullOrEmpty(valueText))
                    {
                        keyValuePairs[keyText] = valueText;
                    }
                }
            }

            return keyValuePairs;
        }

        private List<Dictionary<string, object>> ExtractTables(List<Block> blocks)
        {
            var tables = new List<Dictionary<string, object>>();

            foreach (var block in blocks.Where(b => b.BlockType == BlockType.TABLE))
            {
                var tableData = ParseTableBlock(block, blocks);
                tables.Add(tableData);
            }

            return tables;
        }

        private string GetTextFromRelationships(Block block, List<Block> allBlocks)
        {
            var textParts = new List<string>();

            if (block.Relationships != null)
            {
                foreach (var relationship in block.Relationships.Where(r => r.Type == RelationshipType.CHILD))
                {
                    foreach (var childId in relationship.Ids)
                    {
                        var childBlock = allBlocks.FirstOrDefault(b => b.Id == childId);
                        if (childBlock?.BlockType == BlockType.WORD)
                        {
                            textParts.Add(childBlock.Text ?? "");
                        }
                    }
                }
            }

            return string.Join(" ", textParts);
        }

        private string GetValueFromKeyBlock(Block keyBlock, List<Block> allBlocks)
        {
            // Implementation to get value from key block relationships
            // This would involve traversing the relationships to find the associated value
            return "";
        }

        private Dictionary<string, object> ParseTableBlock(Block tableBlock, List<Block> allBlocks)
        {
            // Implementation to parse table structure
            return new Dictionary<string, object>();
        }

        private List<Dictionary<string, object>> ParseTransactionTables(List<Dictionary<string, object>> tables)
        {
            // Implementation to parse transaction data from tables
            return new List<Dictionary<string, object>>();
        }

        private Dictionary<string, double> CalculateConfidenceScores(List<Block> blocks)
        {
            var confidences = blocks
                .Where(b => b.Confidence.HasValue)
                .Select(b => b.Confidence.Value)
                .ToList();

            if (!confidences.Any())
            {
                return new Dictionary<string, double>
                {
                    ["average"] = 0.0,
                    ["minimum"] = 0.0,
                    ["maximum"] = 0.0
                };
            }

            return new Dictionary<string, double>
            {
                ["average"] = confidences.Average(),
                ["minimum"] = confidences.Min(),
                ["maximum"] = confidences.Max()
            };
        }
    }
}